File: completed/0287.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        unordered_set<int> p;
        for (int num : nums) {
            if (p.find(num) != p.end()) {
                return num;
            }
            p.insert(num);
        }
        return -1;
    }
};

int main() {
    vector<int> nums = { 1, 3, 4, 2, 2 };

    Solution solution;
    cout << solution.findDuplicate(nums) << endl;
    return 0;
}

File: completed/0287.go
package main

import "fmt"

func findDuplicate(nums []int) int {
    slow := nums[nums[0]]
    fast := nums[nums[nums[0]]]
    for slow != fast {
        slow = nums[slow]
        fast = nums[nums[fast]]
    }
    slow = nums[0]
    for slow != fast {
        slow = nums[slow]
        fast = nums[fast]
    }
    return slow
}

func main() {
    nums := []int{3,1,3,4,2}
    fmt.Println(findDuplicate(nums))
}
